core/src/avm1/object/file_reference.rs

```diff
@@ -0,0 +1,119 @@
+use crate::add_field_accessors;
+use crate::avm1::activation::Activation;
+use crate::avm1::object::date_object::DateObject;
+use crate::avm1::{Object, ScriptObject, TObject};
+use crate::backend::ui::FileDialogResult;
+use crate::impl_custom_object;
+use gc_arena::{Collect, GcCell, MutationContext};
+use std::fmt;
+
+/// A FileReference
+#[derive(Clone, Copy, Collect)]
+#[collect(no_drop)]
+pub struct FileReferenceObject<'gc>(GcCell<'gc, FileReferenceData<'gc>>);
+
+#[derive(Clone, Collect)]
+#[collect(no_drop)]
+pub struct FileReferenceData<'gc> {
+    /// The underlying script object.
+    base: ScriptObject<'gc>,
+
+    creation_date: Option<DateObject<'gc>>,
+    creator: GcCell<'gc, Option<String>>,
```

Comment:
do these fields need to be behind `GcCell`s?

Reply:
Well, the custom object macro creates copying read functions. Since String does not copy by default it fails. If I create a custom getter I'll still need to clone the value since I can't return a reference - it is still owned by the overarching FileReferenceObject.

So I can leave as is or create setters and getters where I specifically clone the return value. I'm not sure which is preferable. Or I'm missing something because rust is toying with me.

Reply:
I think we should avoid adding extra GC layers like that (unless we made these AvmStrings, which I'm not sure is better or worse).

Internally this also makes the object slightly simpler to use in some places, like:

```
    pub fn init_from_dialog_result(...)
        let write = self.0.write(activation.context.gc_context);

        write.file_type = dialog_result.file_type();
        write.name = dialog_result.file_name();
        write.size = dialog_result.size();
        write.creator = dialog_result.creator();
        (...)
```

> If I create a custom getter I'll still need to clone the value since I can't return a reference

To avoid copying, I think you can return `-> Ref<str>`, that's what we do in DisplayObject. On the other hand, SharedObject indeed makes a copy for `name`. Actually either way it's not really much of an issue, as when returning an AVM value you always need to clone to move the new string into an `AvmString`.